src/Common/Compiler.hs

```diff
@@ -1,52 +1,68 @@
 {-# LANGUAGE DerivingVia #-}
+-- | Data types and helpers used to compose the compiler pipeline.
 module Common.Compiler
   ( ErrorMsg
   , Error(..)
   , Pass(..)
+  , fromString
   , runPass
   , dump
   , throw
   , passIO
+  , liftEither
   ) where
 
 import           Common.Pretty                  ( Pretty(pretty) )
 import           Control.Monad.Except           ( Except
                                                 , MonadError(..)
                                                 , runExcept
                                                 , throwError
+                                                , liftEither
                                                 )
 
+import           Data.String                    ( IsString(..) )
 import           System.Exit                    ( exitFailure
                                                 , exitSuccess
                                                 )
 import           System.IO                      ( hPrint
                                                 , stderr
                                                 )
 
--- | Convenience
-type ErrorMsg = String
+-- | Type for error messages.
+newtype ErrorMsg = ErrorMsg String
+  deriving Show
+  deriving Semigroup    via String
+  deriving Monoid       via String
+
+instance IsString ErrorMsg where
+  fromString = ErrorMsg
+
+instance Eq ErrorMsg where
+  _ == _ = True
```

Comment:
By overloading `ErrorMsg`, we can treat all error messages as being equal. Then, when we compare equality on `Error`s, we are just comparing the broader category of error. This allows us to check which kind of an error we are throwing without having to restrict the contents of the error message we report.

Correspondingly, variants of `Error` should be chosen according to the granularity of which classes of errors we consider equal.